### PR TITLE
add NgPackagr aspect & support for bit compile & bit build (#4170)

### PR DESCRIPTION
**Pull request by ocombe**
## Proposed Changes

- Add a new ngPackagr aspect to compile Angular code
- Support for build & compile commands (except the preview part in the build command)
- Refactor Environment to add more details about the different functions (Linter, Dev, Builder, ...) <-- @ranm8 David told me that you're not really fan of refactoring this, but I think that it helps a lot when writing a new environment... let me know if you prefer another way (like maybe exporting those extended interfaces from their own packages, like `BuilderEnv` from `@teambit/builder`?)

As usual, this PR is not merged into master but into the harmony/angular branch, to make sure that we don't release an unstable feature yet !

### CODE REVIEW

File: scopes/angular/angular/angular.env.ts [hunk 1]
```diff
@@ -1,65 +1,117 @@
...
+   * Returns the list of dependencies
+   * Required for any task
+   */
   getDependencies() {
     return {
       dependencies: {
         '@angular/common': '-',
         '@angular/core': '-',
-        tslib: '-',
+        tslib: '^2.0.0',
```
Comment (GiladShoham) [inline scopes/angular/angular/angular.env.ts:114]: is this required to be dependency of any component uses the angular env?

File: scopes/angular/angular/angular.main.runtime.ts [hunk 2]
```diff
@@ -2,22 +2,93 @@ import { MainRuntime } from '@teambit/cli';
...
+    NgPackagrAspect,
+  ];
   static runtime = MainRuntime;
-  static async provider([envs, eslint, generator]: AngularDeps) {
-    const angularEnv = new AngularEnv(eslint);
-    const angularMain = new AngularMain(/* reactEnv, envs, application, workspace */);
+
+  constructor(
+    /**
+     * an instance of the React env.
```
Comment (GiladShoham) [inline scopes/angular/angular/angular.main.runtime.ts:53]: react -> angular

File: scopes/angular/angular/templates/angular-component/src/lib/component-spec.ts [hunk 3]
```diff
@@ -1,10 +1,10 @@
 import { ComponentContext } from '@teambit/generator';
 
-export const testFile = (context: ComponentContext) => {
+export const componentSpecFile = (context: ComponentContext) => {
   const { name, namePascalCase: Name } = context;
 
   return {
-    relativePath: `${name}.spec.ts`,
+    relativePath: `src/${name}.spec.ts`,
```
Comment (GiladShoham) [inline scopes/angular/angular/templates/angular-component/src/lib/component-spec.ts:7]: is it necessary to have each file under the `src` folder within the component?

Comment (GiladShoham) [reply scopes/angular/angular/templates/angular-component/src/lib/component-spec.ts:7]: I guess it's related to the public_api file is outside the src?

File: scopes/angular/ng-packagr/component.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      "packageManagerPublishArgs": [
+        "--access public"
+      ]
+    },
+    "teambit.dependencies/dependency-resolver": {
+      "policy": {
+        "devDependencies": {
+          "@angular/compiler": "^11.0.0",
+          "@angular/compiler-cli": "^11.0.0",
+          "ng-packagr": "^11.0.0"
```
Comment (GiladShoham) [inline scopes/angular/ng-packagr/component.json:22]: I would consider locking those to exact versions

File: scopes/angular/ng-packagr/ng-packagr.compiler.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  getDistPathBySrcPath(srcPath: string): string {
+    const fileWithJSExtIfNeeded = this.replaceFileExtToJs(srcPath);
+    return join(this.distDir, fileWithJSExtIfNeeded);
+  }
+
+  /**
+   * whether ngPackagr is able to compile the given path
+   */
+  isFileSupported(filePath: string): boolean {
+    return filePath.endsWith('.ts');
```
Comment (GiladShoham) [inline scopes/angular/ng-packagr/ng-packagr.compiler.ts:147]: js / jsx / tsx are not supported by definition?

File: scopes/compilation/compiler/types.ts [hunk 6]
```diff
@@ -1,11 +1,17 @@
 import { BuildContext, BuildTask, BuiltTaskResult, TaskResultsList } from '@teambit/builder';
+import ConsumerComponent from '@teambit/legacy/dist/consumer/component';
```
Comment (GiladShoham) [inline scopes/compilation/compiler/types.ts:2]: having something from teambit/legacy here is problematic. let's discuss it.

File: scopes/compilation/compiler/types.ts [hunk 7]
```diff
@@ -50,15 +56,23 @@ export interface Compiler extends CompilerOptions {
...
+   * transpile a single file that gets saved into the workspace, used by `bit compile` and during
+   * development
+   */
+  transpileFile?: (fileContent: string, params: TranspileFileParams) => TranspileFileOutput;
+
+  /**
+   * transpile all the files of a component, use this when you can't use `transpileFile`
    */
-  transpileFile: (fileContent: string, options: TranspileOpts) => TranspileOutput;
+  transpileComponent?: (component: ConsumerComponent, params: TranspileComponentParams) => Promise<void>;
```
Comment (GiladShoham) [inline scopes/compilation/compiler/types.ts:75]: no relevant data we can return back here?

File: scopes/envs/envs/environment.ts [hunk 8]
```diff
@@ -1,4 +1,21 @@
...
+import { SchemaExtractor } from '@teambit/schema';
+import { WebpackConfigTransformer } from '@teambit/webpack';
+import { PackageJsonProps } from '@teambit/pkg';
+import { TsConfigSourceFile } from 'typescript';
+
+export type EnvDescriptor = {
+  type: string;
+};
+
+export interface PkgDependencies {
```
Comment (GiladShoham) [inline scopes/envs/envs/environment.ts:14]: This type shouldn't be here. should probably be in the pkg or in dependency resolver.

File: scopes/envs/envs/environment.ts [hunk 9]
```diff
@@ -22,7 +39,87 @@ export interface Environment {
...
-   * Returns the linter to use (ESLint, ...)
+   * @deprecated Fatal: a breaking API has introduced. Use getBuildPipe() instead
+   */
+  getPipe?: () => BuildTask[];
+
+  /**
+   * Returns the list of dependencies
+   * Required for any task
+   */
+  getDependencies?: () => PkgDependencies | Promise<PkgDependencies>;
```
Comment (GiladShoham) [inline scopes/envs/envs/environment.ts:50]: extract to separate interface it's not part of the env's core

File: scopes/envs/envs/environment.ts [hunk 10]
```diff
@@ -22,7 +39,87 @@ export interface Environment {
...
+
+  /**
+   * define the package json properties to add to each component.
+   * Used by `bit link` to augment package.json with new properties
+   */
+  getPackageJsonProps?: () => PackageJsonProps;
```
Comment (GiladShoham) [inline scopes/envs/envs/environment.ts:56]: extract to separate interface it's not part of the env's core

File: scopes/workspace/workspace/workspace.ts [hunk 11]
```diff
@@ -1191,6 +1189,13 @@ export class Workspace implements ComponentFactory {
     this.clearCache();
   }
 
+  getLocalDir(component: ConsumerComponent | Component) {
```
Comment (GiladShoham) [inline scopes/workspace/workspace/workspace.ts:1192]: change the name, this name is confusing. I might see it and think that it will give me the component root dir in the workspace.
This is an external API so the name is important

**Review by GiladShoham — CHANGES_REQUESTED** [review]

File: scopes/angular/angular/angular.env.ts (hunk 1 above)
Comment (ocombe) [reply scopes/angular/angular/angular.env.ts:114]: Yes, otherwise ngPackagr will print a warning that it should be a dependency and not a peer dependency. Not 
sure why though

File: scopes/angular/angular/templates/angular-component/src/lib/component-spec.ts (hunk 3 above)
Comment (ocombe) [reply scopes/angular/angular/templates/angular-component/src/lib/component-spec.ts:7]: yes, I'm just trying to follow the style that Angular promotes (based on their style guide and what angular CLI generates), but if you have a good reason to avoid that I don't really mind

File: scopes/angular/ng-packagr/component.json (hunk 4 above)
Comment (ocombe) [reply scopes/angular/ng-packagr/component.json:22]: Angular works really hard to follow semver and they will not make a breaking change in a minor/patch, I don't think it really matters ?

File: scopes/angular/ng-packagr/ng-packagr.compiler.ts (hunk 5 above)
Comment (ocombe) [reply scopes/angular/ng-packagr/ng-packagr.compiler.ts:147]: jsx/tsx I don't think so, js should work as long as the TS option is enabled, I'll test it

File: scopes/compilation/compiler/types.ts (hunk 7 above)
Comment (ocombe) [reply scopes/compilation/compiler/types.ts:75]: we could at least return errors if any

File: scopes/compilation/compiler/types.ts (hunk 6 above)
Comment (ocombe) [reply scopes/compilation/compiler/types.ts:2]: we can probably use `Consumer` instead, let me check

File: scopes/compilation/compiler/types.ts (hunk 7 above)
Comment (ocombe) [reply scopes/compilation/compiler/types.ts:75]: Maybe a boolean for success ?

File: scopes/envs/envs/environment.ts (hunk 9 above)
Comment (ocombe) [reply scopes/envs/envs/environment.ts:50]: Are you sure? it seems to be used by most tasks

File: scopes/workspace/workspace/workspace.ts (hunk 11 above)
Comment (ocombe) [reply scopes/workspace/workspace/workspace.ts:1192]: Changed the name to `getComponentPackagePath` and made it private as it's only used in workspace now (I can even remove this change if you want)

**Comment by ocombe** [comment]
I've updated the PR, and added support for JS compilation

File: scopes/envs/envs/environment.ts (hunk 9 above)
Comment (GiladShoham) [reply scopes/envs/envs/environment.ts:39]: yes I'm sure:) we can discuss it if you want (both this and the getPackageJsonProps)

**Review by GiladShoham — CHANGES_REQUESTED** [review]
